File: lib/warmup.ts
export type Drill = 'arpeggio' | 'five-note';
export type Beat = {
  time: number;
  midi: number | null;
  root: number;
  round: number;
  label: string;
  accent: boolean;
  phase: 'listen' | 'count' | 'sing' | 'rest';
  noteIndex: number;
  onset: boolean;
};
export function makeWarmup(
  drill: Drill,
  root: number,
  steps: number,
  bpm: number,
): Beat[] {
  if (
    !['arpeggio', 'five-note'].includes(drill) ||
    !Number.isInteger(root) ||
    root < 36 ||
    root > 72 ||
    !Number.isInteger(steps) ||
    steps < 0 ||
    steps > 12 ||
    bpm < 50 ||
    bpm > 140 ||
    !Number.isFinite(bpm)
  )
    throw new Error('Invalid warm-up settings');
  const pattern =
    drill === 'arpeggio' ? [0, 4, 7, 12, 7, 4, 0] : [0, 2, 4, 5, 7, 5, 4, 2, 0];
  const beats: Beat[] = [];
  const duration = 60 / bpm;
  const add = (
    midi: number | null,
    r: number,
    round: number,
    label: string,
    accent: boolean,
    phase: Beat['phase'],
    noteIndex = -1,
    onset = true,
  ) =>
    beats.push({
      time: beats.length * duration,
      midi,
      root: r,
      round,
      label,
      accent,
      phase,
      noteIndex,
      onset,
    });
  for (let n = 0; n < 4; n++)
    add(null, root, 1, `Listen in ${4 - n}`, n === 0, 'count');
  for (let s = 0; s <= steps; s++) {
    for (const phase of ['listen', 'sing'] as const) {
      if (phase === 'sing')
        for (let n = 0; n < 4; n++)
          add(null, root + s, s + 1, `Your turn in ${4 - n}`, n === 0, 'count');
      pattern.forEach((offset, i) => {
        for (let hold = 0; hold < 2; hold++)
          add(
            root + s + offset,
            root + s,
            s + 1,
            phase === 'listen' ? 'Listen to the piano' : 'Your turn · sing',
            hold === 0,
            phase,
            i,
            hold === 0,
          );
      });
    }
    for (let n = 0; n < 4; n++)
      add(null, root + s, s + 1, `Breathe · ${4 - n}`, n === 0, 'rest');
  }
  return beats;
}
export type NoteScore = {
  total: number;
  voiced: number;
  hits: number;
  cents: number[];
};
export function scoreSample(
  previous: NoteScore | undefined,
  midi: number | null,
  target: number,
): NoteScore {
  const s = previous || { total: 0, voiced: 0, hits: 0, cents: [] };
  const diff = midi === null ? null : (midi - target) * 100;
  return {
    total: s.total + 1,
    voiced: s.voiced + (diff === null ? 0 : 1),
    hits: s.hits + (diff !== null && Math.abs(diff) <= 50 ? 1 : 0),
    cents: diff === null ? s.cents : [...s.cents, diff],
  };
}
export function scoreLabel(s: NoteScore | undefined): string {
  if (!s || s.voiced < 3) return 'No clear pitch';
  if (s.hits >= 3 && s.hits / s.total >= 0.5) return 'Hit';
  const sorted = [...s.cents].sort((a, b) => a - b);
  const median = sorted[Math.floor(sorted.length / 2)];
  return median < -50 ? 'Low' : median > 50 ? 'High' : 'Keep steady';
}
export const formatTime = (seconds: number) =>
  `${Math.floor(seconds / 60)}:${String(Math.floor(seconds % 60)).padStart(2, '0')}`;

File: ios/Packages/SingwellCore/Sources/SingwellCore/Warmup.swift
import Foundation

public enum Warmup {
    public enum Drill: String, CaseIterable, Codable, Sendable, Identifiable {
        case arpeggio
        case fiveNote = "five-note"
        public var id: String { rawValue }
        public var label: String {
            switch self {
            case .arpeggio: return "Octave arpeggio"
            case .fiveNote: return "Five-note scale"
            }
        }
        var pattern: [Int] {
            switch self {
            case .arpeggio: return [0, 4, 7, 12, 7, 4, 0]
            case .fiveNote: return [0, 2, 4, 5, 7, 5, 4, 2, 0]
            }
        }
    }

    public enum Phase: String, Sendable { case listen, count, sing, rest }

    public struct Beat: Equatable, Sendable {
        public var time: Double
        public var midi: Int?
        public var root: Int
        public var round: Int
        public var label: String
        public var accent: Bool
        public var phase: Phase
        public var noteIndex: Int
        public var onset: Bool
    }

    public enum SettingsError: Error { case invalid }

    public static let rootRange = 36...72
    public static let stepsRange = 0...12
    public static let bpmRange = 50.0...140.0

    /// Piano demonstration → count-in → sung pass → breathe → up a semitone.
    public static func plan(drill: Drill, root: Int, steps: Int, bpm: Double) throws -> [Beat] {
        guard rootRange.contains(root), stepsRange.contains(steps), bpm.isFinite, bpmRange.contains(bpm) else {
            throw SettingsError.invalid
        }
        let pattern = drill.pattern
        var beats: [Beat] = []
        let duration = 60 / bpm
        func add(_ midi: Int?, _ r: Int, _ round: Int, _ label: String, _ accent: Bool, _ phase: Phase, _ noteIndex: Int = -1, _ onset: Bool = true) {
            beats.append(Beat(time: Double(beats.count) * duration, midi: midi, root: r, round: round, label: label, accent: accent, phase: phase, noteIndex: noteIndex, onset: onset))
        }
        for n in 0..<4 { add(nil, root, 1, "Listen in \(4 - n)", n == 0, .count) }
        for s in 0...steps {
            for phase in [Phase.listen, Phase.sing] {
                if phase == .sing {
                    for n in 0..<4 { add(nil, root + s, s + 1, "Your turn in \(4 - n)", n == 0, .count) }
                }
                for (i, offset) in pattern.enumerated() {
                    for hold in 0..<2 {
                        add(root + s + offset, root + s, s + 1,
                            phase == .listen ? "Listen to the piano" : "Your turn · sing",
                            hold == 0, phase, i, hold == 0)
                    }
                }
            }
            for n in 0..<4 { add(nil, root + s, s + 1, "Breathe · \(4 - n)", n == 0, .rest) }
        }
        return beats
    }

    public struct NoteScore: Equatable, Sendable {
        public var total = 0
        public var voiced = 0
        public var hits = 0
        public var cents: [Double] = []
        public init() {}
    }

    public static func score(_ previous: NoteScore?, midi: Double?, target: Int) -> NoteScore {
        var s = previous ?? NoteScore()
        let diff: Double? = midi.map { ($0 - Double(target)) * 100 }
        s.total += 1
        if let diff {
            s.voiced += 1
            if abs(diff) <= 50 { s.hits += 1 }
            s.cents.append(diff)
        }
        return s
    }

    public enum Verdict: String, Sendable {
        case hit = "Hit", low = "Low", high = "High", steady = "Keep steady", none = "No clear pitch"
    }

    public static func verdict(_ s: NoteScore?) -> Verdict {
        guard let s, s.voiced >= 3 else { return .none }
        if s.hits >= 3 && Double(s.hits) / Double(s.total) >= 0.5 { return .hit }
        let sorted = s.cents.sorted()
        let median = sorted[sorted.count / 2]
        return median < -50 ? .low : median > 50 ? .high : .steady
    }
}
